add a new interface to query format feature info

yiwenxue opened the pull request:
Re: cocos-creator/3d-tasks#8203

Native engine: https://github.com/cocos-creator/engine-native/pull/4169#issue-1101229496

Move the query of format feature information to the new interface. In the old method, one can only know whether the device generally supports a feature, but the fact is that some formats support this feature, and others don't. With the new interface, one can query whether a feature is supported by a specific format or not.

yiwenxue committed:
add format feature information for query

yiwenxue committed:
adapted the higher level code to use the new formatFeature query inferface, and run the test-cases-3d

yiwenxue committed:
remove uncessary public interface

yiwenxue committed:
remove uncessary public interface

yiwenxue committed:
remove unnecessary api

yiwenxue committed:
fixed the name of function support16FTexture.

YunHsiao commented on cocos/3d/assets/morph-rendering.ts:
```diff
@@ -473,7 +476,7 @@ class MorphUniforms {
  */
 function createVec4TextureFactory (gfxDevice: Device, vec4Capacity: number) {
-    const hasFeatureFloatTexture = gfxDevice.hasFeature(Feature.TEXTURE_FLOAT);
+    const hasFeatureFloatTexture = gfxDevice.getFormatFeature(Format.RGBA32F) & (FormatFeatureBit.RENDER_TARGET | FormatFeatureBit.SAMPLED_TEXTURE);
```
Morph rendering doesn't require the RT flag to be set, only sampled texture feature would be suffice

YunHsiao commented on cocos/3d/skeletal-animation/skeletal-animation-utils.ts:
```diff
@@ -47,7 +47,7 @@ export const uploadJointData = uploadJointDataLBS;
 
 export function selectJointsMediumFormat (device: Device): Format {
-    if (device.hasFeature(Feature.TEXTURE_FLOAT)) {
+    if (device.getFormatFeature(Format.RGBA32F) & FormatFeatureBit.RENDER_TARGET) {
```
here too, only test sampled texture feature

YunHsiao commented on cocos/core/gfx/webgl/webgl-commands.ts:
```diff
@@ -826,7 +826,9 @@ export function WebGLCmdFuncCreateTexture (device: WebGLDevice, gpuTexture: IWeb
 
-        if (!device.extensions.WEBGL_depth_texture && FormatInfos[gpuTexture.format].hasDepth) {
+        // @ts-expect-error(2341) friend class access
+        const { _textureExclusive: textureExclusive } = device;
```
Just add a public getter for the property (since it's child-class-only), don't need to do this here

YunHsiao commented on cocos/core/gfx/webgl/webgl-device.ts:
```diff
@@ -245,6 +213,166 @@ export class WebGLDevice extends Device {
+            this._textureExclusive[Format.SRGB8_A8] = false;
+        }
+
+        if (exts.WEBGL_depth_texture) {
```
if this extension is not present, then these two format won't be able to be sampled from (renderbuffer only)

YunHsiao replied:
if this extension is not present, then these two format won't be able to be sampled from (renderbuffer only)

YunHsiao commented on cocos/core/gfx/webgl/webgl-device.ts:
```diff
@@ -245,6 +213,166 @@ export class WebGLDevice extends Device {
+
+        if (exts.EXT_color_buffer_half_float) {
+            this._textureExclusive[Format.RGB16F] = false;
+            this._textureExclusive[Format.RGBA16F] = false;
```
should enable the RT feature here

YunHsiao commented on cocos/core/gfx/webgl/webgl-device.ts:
```diff
@@ -245,6 +213,166 @@ export class WebGLDevice extends Device {
+        if (exts.OES_texture_float) {
+            tempFeature = FormatFeatureBit.RENDER_TARGET | FormatFeatureBit.SAMPLED_TEXTURE;
+            this._formatFeatures[Format.RGB32F] = tempFeature;
+            this._formatFeatures[Format.RGBA32F] = tempFeature;
```
should add the sampled texture feature on top of previous features here

YunHsiao commented on cocos/core/pipeline/define.ts:
```diff
@@ -594,20 +596,16 @@ export const MODEL_ALWAYS_MASK = Layers.Enum.ALL;
 /**
- * @en Does the device support half float texture? (for both color attachment and sampling)
- * @zh 当前设备是否支持半浮点贴图？（颜色输出和采样）
+ * @en Does the device support float texture? (for both color attachment and sampling)
```
single-channeled float texture

YunHsiao commented:
The bitwise manipulation in the declaration process should also be changed to the same paradigm as mentioned in the native PR

yiwenxue committed:
clean the code, fix some typos.

yiwenxue committed:
correct format features

yiwenxue committed:
group the judgemnts to fix the logic

YunHsiao commented on cocos/core/gfx/webgl2/webgl2-device.ts:
```diff
@@ -140,53 +145,33 @@ export class WebGL2Device extends Device {
         let compressedFormat = '';
 
         if (exts.WEBGL_compressed_texture_etc1) {
-            this._features[Feature.FORMAT_ETC1] = true;
             compressedFormat += 'etc1 ';
```
```suggestion
        if (this.getFormatFeatures(Format.ETC_RGB8)) {
            compressedFormat += 'etc1 ';
```

YunHsiao replied:
Would be better to unify all these conditions to use the result from format feature interface instead of doing it all over again. This applies to the native engine as well.

yiwenxue committed:
query the support to compressed from a decent method. correct some bit operations.

yiwenxue committed:
Merge branch 'v3.5.0' into issue8203

minggo commented:
npm test breaks.

yiwenxue committed:
add FormatFeatureBit into globaleVariable's database to pass npm test

pandamicro commented on cocos/core/gfx/base/define.ts:
```diff
@@ -103,20 +103,6 @@ export enum SurfaceTransform {
-    FORMAT_DXT,
-    FORMAT_PVRTC,
-    FORMAT_ASTC,
-    FORMAT_RGB8,
```
This looks like breaking changes, if user have used these Feature flags in previous version, what will happen in the new version ?

YunHsiao replied:
Yes, we can add some deprecations for this

pandamicro commented on tests/general/sealed-global-variables.json:
```diff
@@ -106,6 +106,7 @@
         "cc.GFXFeature",
         "cc.GFXFilter",
         "cc.GFXFormat",
+        "cc.GFXFormatFeatureBit",
```
I didn't see this have been exported anywhere, where have the global variable been assigned

YunHsiao replied:
These are auto-generated by deprecation

pandamicro commented on tests/general/sealed-global-variables.json:
```diff
@@ -106,6 +106,7 @@
         "cc.GFXFeature",
         "cc.GFXFilter",
         "cc.GFXFormat",
+        "cc.GFXFormatFeatureBit",
```
I didn't see this have been exported anywhere, where have the global variable been assigned

yiwenxue committed:
Add deprecation information to help developers move to the new
interface. Every time a deprecated property is accessed, an error will
present. So developers must move to the new interface following the
instruction in the error info.

yiwenxue committed:
Merge branch 'v3.5.0' into issue8203

yiwenxue committed:
Move the deprecation information to the right module.
